{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook details the parsing of mouse phenotype ontology gene sets from Mouse Genome Informatics (MGI)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The relevant files to curate the mouse gene set were downloaded from MGI.\n",
    "\n",
    "http://www.informatics.jax.org/downloads/reports/index.html\n",
    "\n",
    "Two files are required:\n",
    "1. MGI_PhenoGenoMP.rpt\n",
    "2. HMD_HumanPhenotype.rpt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_file = \"../data/gene_sets_June2022/MGI_PhenoGenoMP.rpt\"\n",
    "human_file = \"../data/gene_sets_June2022/HMD_HumanPhenotype.rpt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing of mouse file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first parse the MGI_PhenoGenoMP.rpt file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allelic_Composition</th>\n",
       "      <th>AlleleSymbol</th>\n",
       "      <th>Genetic_Background</th>\n",
       "      <th>Mammalian_Phenotype_ID</th>\n",
       "      <th>PubMed_ID</th>\n",
       "      <th>MGI_Marker_Accession_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000600</td>\n",
       "      <td>12529408</td>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001716</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001698</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0001092</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;/Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>Rb1&lt;tm1Tyj&gt;</td>\n",
       "      <td>involves: 129S2/SvPas</td>\n",
       "      <td>MP:0000961</td>\n",
       "      <td>16449662</td>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Allelic_Composition AlleleSymbol     Genetic_Background  \\\n",
       "0  Rb1<tm1Tyj>/Rb1<tm1Tyj>  Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "1  Rb1<tm1Tyj>/Rb1<tm1Tyj>  Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "2  Rb1<tm1Tyj>/Rb1<tm1Tyj>  Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "3  Rb1<tm1Tyj>/Rb1<tm1Tyj>  Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "4  Rb1<tm1Tyj>/Rb1<tm1Tyj>  Rb1<tm1Tyj>  involves: 129S2/SvPas   \n",
       "\n",
       "  Mammalian_Phenotype_ID PubMed_ID MGI_Marker_Accession_ID  \n",
       "0             MP:0000600  12529408               MGI:97874  \n",
       "1             MP:0001716  16449662               MGI:97874  \n",
       "2             MP:0001698  16449662               MGI:97874  \n",
       "3             MP:0001092  16449662               MGI:97874  \n",
       "4             MP:0000961  16449662               MGI:97874  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse = pd.read_csv(mouse_file,header=None,sep='\\t')\n",
    "mouse.columns = [\"Allelic_Composition\", \"AlleleSymbol\",\n",
    "                 \"Genetic_Background\",\"Mammalian_Phenotype_ID\",\n",
    "                 \"PubMed_ID\",\"MGI_Marker_Accession_ID\"]\n",
    "mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allelic_Composition</th>\n",
       "      <th>AlleleSymbol</th>\n",
       "      <th>Genetic_Background</th>\n",
       "      <th>Mammalian_Phenotype_ID</th>\n",
       "      <th>PubMed_ID</th>\n",
       "      <th>MGI_Marker_Accession_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>358233</th>\n",
       "      <td>Rr27&lt;tm1Kpfe&gt;/Rr27&lt;+&gt;,Tg(Tpbpa-cre,-EGFP)5Jcc/0</td>\n",
       "      <td>Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27&lt;tm1Kpfe&gt;|Rr27&lt;+&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ</td>\n",
       "      <td>MP:0005033</td>\n",
       "      <td>34982814</td>\n",
       "      <td>MGI:5287871|MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358234</th>\n",
       "      <td>Rr27&lt;tm1Kpfe&gt;/Rr27&lt;+&gt;,Tg(Tpbpa-cre,-EGFP)5Jcc/0</td>\n",
       "      <td>Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27&lt;tm1Kpfe&gt;|Rr27&lt;+&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ</td>\n",
       "      <td>MP:0004256</td>\n",
       "      <td>34982814</td>\n",
       "      <td>MGI:5287871|MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358235</th>\n",
       "      <td>Rr27&lt;tm1Kpfe&gt;/Rr27&lt;+&gt;,Tg(Tpbpa-cre,-EGFP)5Jcc/0</td>\n",
       "      <td>Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27&lt;tm1Kpfe&gt;|Rr27&lt;+&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ</td>\n",
       "      <td>MP:0008957</td>\n",
       "      <td>34982814</td>\n",
       "      <td>MGI:5287871|MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358236</th>\n",
       "      <td>Rr27&lt;tm1Kpfe&gt;/Rr27&lt;+&gt;,Tg(Tpbpa-cre,-EGFP)5Jcc/0</td>\n",
       "      <td>Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27&lt;tm1Kpfe&gt;|Rr27&lt;+&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ</td>\n",
       "      <td>MP:0008959</td>\n",
       "      <td>34982814</td>\n",
       "      <td>MGI:5287871|MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358237</th>\n",
       "      <td>Rr27&lt;tm1Kpfe&gt;/Rr27&lt;+&gt;,Tg(Tpbpa-cre,-EGFP)5Jcc/0</td>\n",
       "      <td>Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27&lt;tm1Kpfe&gt;|Rr27&lt;+&gt;</td>\n",
       "      <td>involves: 129S1/Sv * 129X1/SvJ</td>\n",
       "      <td>MP:0011521</td>\n",
       "      <td>34982814</td>\n",
       "      <td>MGI:5287871|MGI:5489785</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Allelic_Composition  \\\n",
       "358233  Rr27<tm1Kpfe>/Rr27<+>,Tg(Tpbpa-cre,-EGFP)5Jcc/0   \n",
       "358234  Rr27<tm1Kpfe>/Rr27<+>,Tg(Tpbpa-cre,-EGFP)5Jcc/0   \n",
       "358235  Rr27<tm1Kpfe>/Rr27<+>,Tg(Tpbpa-cre,-EGFP)5Jcc/0   \n",
       "358236  Rr27<tm1Kpfe>/Rr27<+>,Tg(Tpbpa-cre,-EGFP)5Jcc/0   \n",
       "358237  Rr27<tm1Kpfe>/Rr27<+>,Tg(Tpbpa-cre,-EGFP)5Jcc/0   \n",
       "\n",
       "                                         AlleleSymbol  \\\n",
       "358233  Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27<tm1Kpfe>|Rr27<+>   \n",
       "358234  Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27<tm1Kpfe>|Rr27<+>   \n",
       "358235  Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27<tm1Kpfe>|Rr27<+>   \n",
       "358236  Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27<tm1Kpfe>|Rr27<+>   \n",
       "358237  Tg(Tpbpa-cre,-EGFP)5Jcc|Rr27<tm1Kpfe>|Rr27<+>   \n",
       "\n",
       "                    Genetic_Background Mammalian_Phenotype_ID PubMed_ID  \\\n",
       "358233  involves: 129S1/Sv * 129X1/SvJ             MP:0005033  34982814   \n",
       "358234  involves: 129S1/Sv * 129X1/SvJ             MP:0004256  34982814   \n",
       "358235  involves: 129S1/Sv * 129X1/SvJ             MP:0008957  34982814   \n",
       "358236  involves: 129S1/Sv * 129X1/SvJ             MP:0008959  34982814   \n",
       "358237  involves: 129S1/Sv * 129X1/SvJ             MP:0011521  34982814   \n",
       "\n",
       "        MGI_Marker_Accession_ID  \n",
       "358233  MGI:5287871|MGI:5489785  \n",
       "358234  MGI:5287871|MGI:5489785  \n",
       "358235  MGI:5287871|MGI:5489785  \n",
       "358236  MGI:5287871|MGI:5489785  \n",
       "358237  MGI:5287871|MGI:5489785  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We notice that the \"MGI_Marker_Accession_ID\" column maps to more than one Mammalian_Phenotype_ID. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI_Marker_Accession_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:97874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358235</th>\n",
       "      <td>MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358236</th>\n",
       "      <td>MGI:5287871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358236</th>\n",
       "      <td>MGI:5489785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358237</th>\n",
       "      <td>MGI:5287871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358237</th>\n",
       "      <td>MGI:5489785</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>449267 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MGI_Marker_Accession_ID\n",
       "0                    MGI:97874\n",
       "1                    MGI:97874\n",
       "2                    MGI:97874\n",
       "3                    MGI:97874\n",
       "4                    MGI:97874\n",
       "...                        ...\n",
       "358235             MGI:5489785\n",
       "358236             MGI:5287871\n",
       "358236             MGI:5489785\n",
       "358237             MGI:5287871\n",
       "358237             MGI:5489785\n",
       "\n",
       "[449267 rows x 1 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some rows have more than one MGI_Marker_Accession_ID\n",
    "# explode columns\n",
    "mouse = mouse[[\"MGI_Marker_Accession_ID\"]].copy()\n",
    "mouse[\"MGI_Marker_Accession_ID\"] = mouse[\"MGI_Marker_Accession_ID\"].apply(lambda x: x.split(\"|\"))\n",
    "mouse = mouse.explode(\"MGI_Marker_Accession_ID\")\n",
    "mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The unique list of MGI ids are outputted to a file and passed through a batch search.\n",
    "\n",
    "http://www.informatics.jax.org/batch\n",
    "\n",
    "Option Mammalian Phenotype (MP) was selected in addition to all gene attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22428"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#from quick visualization, we see that the MGI_Marker_Accession_ID is not unique\n",
    "# mouse_mgi_ids_parsed\n",
    "mouse = mouse.drop_duplicates(\"MGI_Marker_Accession_ID\")\n",
    "mouse[\"MGI_Marker_Accession_ID\"].to_csv(\"../data/gene_sets_June2022/MGI_ID_unique_27June_2022.txt\",\n",
    "                                        index=None, header=False)\n",
    "len(mouse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing of batch report file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>Input Type</th>\n",
       "      <th>MGI Gene/Marker ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ensembl ID</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>MP ID</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB transcriptional corepressor 1</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>73430298.0</td>\n",
       "      <td>73563446.0</td>\n",
       "      <td>ENSMUSG00000022105</td>\n",
       "      <td>19645.0</td>\n",
       "      <td>MP:0000042</td>\n",
       "      <td>abnormal organ of Corti morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB transcriptional corepressor 1</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>73430298.0</td>\n",
       "      <td>73563446.0</td>\n",
       "      <td>ENSMUSG00000022105</td>\n",
       "      <td>19645.0</td>\n",
       "      <td>MP:0000160</td>\n",
       "      <td>kyphosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB transcriptional corepressor 1</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>73430298.0</td>\n",
       "      <td>73563446.0</td>\n",
       "      <td>ENSMUSG00000022105</td>\n",
       "      <td>19645.0</td>\n",
       "      <td>MP:0000163</td>\n",
       "      <td>abnormal cartilage morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB transcriptional corepressor 1</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>73430298.0</td>\n",
       "      <td>73563446.0</td>\n",
       "      <td>ENSMUSG00000022105</td>\n",
       "      <td>19645.0</td>\n",
       "      <td>MP:0000245</td>\n",
       "      <td>abnormal erythropoiesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:97874</td>\n",
       "      <td>Rb1</td>\n",
       "      <td>RB transcriptional corepressor 1</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>73430298.0</td>\n",
       "      <td>73563446.0</td>\n",
       "      <td>ENSMUSG00000022105</td>\n",
       "      <td>19645.0</td>\n",
       "      <td>MP:0000350</td>\n",
       "      <td>abnormal cell proliferation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Input Input Type MGI Gene/Marker ID Symbol  \\\n",
       "0  MGI:97874        MGI          MGI:97874    Rb1   \n",
       "1  MGI:97874        MGI          MGI:97874    Rb1   \n",
       "2  MGI:97874        MGI          MGI:97874    Rb1   \n",
       "3  MGI:97874        MGI          MGI:97874    Rb1   \n",
       "4  MGI:97874        MGI          MGI:97874    Rb1   \n",
       "\n",
       "                               Name         Feature Type Chr Strand  \\\n",
       "0  RB transcriptional corepressor 1  protein coding gene  14      -   \n",
       "1  RB transcriptional corepressor 1  protein coding gene  14      -   \n",
       "2  RB transcriptional corepressor 1  protein coding gene  14      -   \n",
       "3  RB transcriptional corepressor 1  protein coding gene  14      -   \n",
       "4  RB transcriptional corepressor 1  protein coding gene  14      -   \n",
       "\n",
       "        Start         End          Ensembl ID  Entrez Gene ID       MP ID  \\\n",
       "0  73430298.0  73563446.0  ENSMUSG00000022105         19645.0  MP:0000042   \n",
       "1  73430298.0  73563446.0  ENSMUSG00000022105         19645.0  MP:0000160   \n",
       "2  73430298.0  73563446.0  ENSMUSG00000022105         19645.0  MP:0000163   \n",
       "3  73430298.0  73563446.0  ENSMUSG00000022105         19645.0  MP:0000245   \n",
       "4  73430298.0  73563446.0  ENSMUSG00000022105         19645.0  MP:0000350   \n",
       "\n",
       "                                 Term  \n",
       "0  abnormal organ of Corti morphology  \n",
       "1                            kyphosis  \n",
       "2       abnormal cartilage morphology  \n",
       "3             abnormal erythropoiesis  \n",
       "4         abnormal cell proliferation  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mgi_batch_report = pd.read_csv(\"../data/gene_sets_June2022/MGIBatchReport_27June_2022.txt\",sep=\"\\t\")\n",
    "mgi_batch_report.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "253312\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>gene_set</th>\n",
       "      <th>mgi</th>\n",
       "      <th>mp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>183745</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MP:0000003_abnormal_adipose_tissue_morphology</td>\n",
       "      <td>MGI:2177469</td>\n",
       "      <td>MP:0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41139</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MP:0000003_abnormal_adipose_tissue_morphology</td>\n",
       "      <td>MGI:895149</td>\n",
       "      <td>MP:0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31672</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MP:0000003_abnormal_adipose_tissue_morphology</td>\n",
       "      <td>MGI:109583</td>\n",
       "      <td>MP:0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159705</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MP:0000003_abnormal_adipose_tissue_morphology</td>\n",
       "      <td>MGI:3612271</td>\n",
       "      <td>MP:0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48307</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MP:0000003_abnormal_adipose_tissue_morphology</td>\n",
       "      <td>MGI:96820</td>\n",
       "      <td>MP:0000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       database                                       gene_set          mgi  \\\n",
       "183745      MGI  MP:0000003_abnormal_adipose_tissue_morphology  MGI:2177469   \n",
       "41139       MGI  MP:0000003_abnormal_adipose_tissue_morphology   MGI:895149   \n",
       "31672       MGI  MP:0000003_abnormal_adipose_tissue_morphology   MGI:109583   \n",
       "159705      MGI  MP:0000003_abnormal_adipose_tissue_morphology  MGI:3612271   \n",
       "48307       MGI  MP:0000003_abnormal_adipose_tissue_morphology    MGI:96820   \n",
       "\n",
       "                mp  \n",
       "183745  MP:0000003  \n",
       "41139   MP:0000003  \n",
       "31672   MP:0000003  \n",
       "159705  MP:0000003  \n",
       "48307   MP:0000003  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select relevant columns\n",
    "mgi_batch_report_parsed = mgi_batch_report[[\"MP ID\",\"Term\", \"MGI Gene/Marker ID\"]].copy()\n",
    "\n",
    "# replace spaces with underscores\n",
    "mgi_batch_report_parsed[\"Term\"] = mgi_batch_report_parsed[\"Term\"].str.replace(\" \",\"_\")\n",
    "\n",
    "# genesets are MP ID + Term\n",
    "mgi_batch_report_parsed[\"gene_set\"] = mgi_batch_report_parsed[\"MP ID\"]+\"_\"+mgi_batch_report_parsed[\"Term\"]\n",
    "mgi_batch_report_parsed = mgi_batch_report_parsed.sort_values(\"gene_set\")\n",
    "mgi_batch_report_parsed.columns = [\"mp\",\"term\",\"mgi\", \"gene_set\"]\n",
    "mgi_batch_report_parsed[\"database\"] = \"MGI\"\n",
    "mgi_batch_report_parsed = mgi_batch_report_parsed[[\"database\",\"gene_set\", \"mgi\", \"mp\"]]\n",
    "\n",
    "# remove any NAs\n",
    "mgi_batch_report_parsed = mgi_batch_report_parsed.dropna()\n",
    "print(len(mgi_batch_report_parsed))\n",
    "mgi_batch_report_parsed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10554\n",
      "20614\n"
     ]
    }
   ],
   "source": [
    "print(len(mgi_batch_report_parsed[\"mp\"].unique()))\n",
    "print(len(mgi_batch_report_parsed[\"mgi\"].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing of human file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have our gene sets, we must parse the human file so that we can merge on MGI to identify the assoicated human gene for a given MP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>EntrezID</th>\n",
       "      <th>mouse_gene_symbol</th>\n",
       "      <th>mgi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>MGI:2152878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>MGI:1917115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>2</td>\n",
       "      <td>A2m</td>\n",
       "      <td>MGI:2449119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>127550</td>\n",
       "      <td>A3galt2</td>\n",
       "      <td>MGI:2685279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>53947</td>\n",
       "      <td>A4galt</td>\n",
       "      <td>MGI:3512453</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  EntrezID mouse_gene_symbol          mgi\n",
       "0        A1BG         1              A1bg  MGI:2152878\n",
       "1        A1CF     29974              A1cf  MGI:1917115\n",
       "2         A2M         2               A2m  MGI:2449119\n",
       "3     A3GALT2    127550           A3galt2  MGI:2685279\n",
       "4      A4GALT     53947            A4galt  MGI:3512453"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human = pd.read_csv(\"../data/gene_sets_June2022/HMD_HumanPhenotype.rpt\", \n",
    "                    sep=\"\\t\", header=None,\n",
    "                    names = [\"gene_symbol\", \"EntrezID\", \n",
    "                             \"mouse_gene_symbol\", \"mgi\", \n",
    "                             \"mp_id_high_level\", \"blank\"]\n",
    "                   )\n",
    "human = human.drop([\"blank\", \"mp_id_high_level\"], axis=1)\n",
    "human.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now merge the two dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>EntrezID</th>\n",
       "      <th>gene_set</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>MP:0000352_decreased_cell_proliferation</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>MP:0001552_increased_circulating_triglyceride_...</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>MP:0002139_abnormal_hepatobiliary_system_physi...</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>MP:0002727_decreased_circulating_insulin_level</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>29974</td>\n",
       "      <td>MP:0002795_dilated_cardiomyopathy</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  EntrezID                                           gene_set  \\\n",
       "0        A1CF     29974            MP:0000352_decreased_cell_proliferation   \n",
       "1        A1CF     29974  MP:0001552_increased_circulating_triglyceride_...   \n",
       "2        A1CF     29974  MP:0002139_abnormal_hepatobiliary_system_physi...   \n",
       "3        A1CF     29974     MP:0002727_decreased_circulating_insulin_level   \n",
       "4        A1CF     29974                  MP:0002795_dilated_cardiomyopathy   \n",
       "\n",
       "  database  \n",
       "0      MGI  \n",
       "1      MGI  \n",
       "2      MGI  \n",
       "3      MGI  \n",
       "4      MGI  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_mouse = pd.merge(human,mgi_batch_report_parsed,on=[\"mgi\"])\n",
    "human_mouse = human_mouse[[\"gene_symbol\",\"EntrezID\",\"gene_set\"]]\n",
    "human_mouse[\"database\"] = \"MGI\"\n",
    "human_mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EntrezID</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000256069</td>\n",
       "      <td>A2MP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000171428</td>\n",
       "      <td>NAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>ENSG00000156006</td>\n",
       "      <td>NAT2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   EntrezID  ensembl_gene_id gene_symbol\n",
       "0         1  ENSG00000121410        A1BG\n",
       "1         2  ENSG00000175899         A2M\n",
       "2         3  ENSG00000256069       A2MP1\n",
       "3         9  ENSG00000171428        NAT1\n",
       "4        10  ENSG00000156006        NAT2"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df = pd.read_csv(\"../data/entrez_ensembl_mapping_v26_16Jun_2022.txt\", sep=\"\\t\")\n",
    "mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = pd.merge(human_mouse, mapping_df, \n",
    "                    on=[\"EntrezID\", \"gene_symbol\"])\n",
    "final_df = final_df.sort_values([\"gene_set\"])\n",
    "final_df = final_df[[\"database\", \"gene_set\", \"ensembl_gene_id\",\"EntrezID\"]]\n",
    "final_df.to_csv(\"../data/gene_sets_June2022/MGI_genesets.txt\",\n",
    "                sep=\"\\t\", header=True, index=None\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
